azure/services/resourcehealth/resourcehealth.go

```diff
@@ -0,0 +1,139 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcehealth
+
+import (
+	"context"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/resourcehealth/mgmt/2020-05-01/resourcehealth"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/conditions"
+)
+
+const serviceName = "resourcehealth"
+
+// ResourceHealthScope defines the scope interface for a resourcehealth service.
+type ResourceHealthScope interface {
+	azure.Authorizer
+	AvailabilityStatusResourceURI() string
+	AvailabilityStatusResource() conditions.Setter
+}
+
+// AvailabilityStatusFilterer transforms the condition derived from the
+// availability status to allow the condition to be overridden in specific
+// circumstances.
+type AvailabilityStatusFilterer interface {
+	AvailabilityStatusFilter(cond *clusterv1.Condition) *clusterv1.Condition
+}
+
+// Service provides operations on Azure resources.
+type Service struct {
+	Scope ResourceHealthScope
+	client
+}
+
+// New creates a new service.
+func New(scope ResourceHealthScope) *Service {
+	return &Service{
+		Scope:  scope,
+		client: newClient(scope),
+	}
+}
+
+// Name returns the service name.
+func (s *Service) Name() string {
+	return serviceName
+}
+
+// Reconcile ensures the resource's availability status is reflected in its own status.
+func (s *Service) Reconcile(ctx context.Context) error {
+	ctx, log, done := tele.StartSpanWithLogger(ctx, "resourcehealth.Service.Reconcile")
+	defer done()
+
+	resource := s.Scope.AvailabilityStatusResourceURI()
+	avail, err := s.GetByResource(ctx, resource)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get availability status for resource %s", resource)
+	}
+	log.V(2).Info("got availability status for resource", "resource", resource, "status", avail)
+
+	cond := azureAvailabilityStatusToCondition(avail)
+	if filterer, ok := s.Scope.(AvailabilityStatusFilterer); ok {
+		cond = filterer.AvailabilityStatusFilter(cond)
+	}
+
+	conditions.Set(s.Scope.AvailabilityStatusResource(), cond)
+
+	if cond.Status == corev1.ConditionFalse {
+		return errors.Errorf("resource is not available: %s", cond.Message)
+	}
+
+	return nil
+}
+
+// Delete is a no-op.
+func (s *Service) Delete(ctx context.Context) error {
+	_, _, done := tele.StartSpanWithLogger(ctx, "resourcehealth.Service.Delete")
+	defer done()
+
+	return nil
+}
+
+// IsManaged always returns true.
+func (s *Service) IsManaged(ctx context.Context) (bool, error) {
+	return true, nil
+}
+
+func azureAvailabilityStatusToCondition(avail resourcehealth.AvailabilityStatus) *clusterv1.Condition {
```

Comment:
this helper could probably go in the converters package

Reply:
Done!